### PR TITLE
fix: make service tests less flaky

### PR DESCRIPTION
**Pull request by craigpastro**
## Description

Make service tests less flaky by:
- Using retryablehttp client everywhere
- Turning off HTTP server when not needed
- Hit the actually HTTP health endpoint when the test requires it

## Review Checklist
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [ ] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: pkg/cmd/service/service_test.go
Comment (elbuo8) [inline pkg/cmd/service/service_test.go:617]: ?

File: pkg/cmd/service/service_test.go
Comment (elbuo8) [inline pkg/cmd/service/service_test.go:565]: This should be happening within `ensureServiceUp` otherwise you'd be copy pasting everywhere.

File: pkg/cmd/service/service_test.go
Comment (elbuo8) [inline pkg/cmd/service/service_test.go:725]: Can we move this within `ensureServiceUp`?

File: pkg/cmd/service/service_test.go
Comment (elbuo8) [inline pkg/cmd/service/service_test.go:680]: 👍🏽 good catch.

Comment (elbuo8) [reply pkg/cmd/service/service_test.go:680]: Come to think of it, why are we `canceling` if we are actually gracefully closing right after? Seems a bit of a mix up there.

**Review by jon-whit — CHANGES_REQUESTED** [review]
Let's talk more about these before merging. I think the number 1 issue with flaky tests is largely due to tests overlapping on the same port. At least this is true in my experience with this project so far. In `cmd/service/service_test.go` we spin up servers using the same port mappings. We should be auto assigning a port using an address such as `:0` or `127.0.0.1:` (https://pkg.go.dev/net#Listen).

I think we should refactor our service tests to auto assign a port dynamically and run the server using that port. This will require some refactoring a bit to make that composition neat and tidy in our tests, but I think it is well worth it.

**Comment by elbuo8** [comment]
@jon-whit agreed. Particularly if want them to run concurrently. But don't think these changes should be blocked until that lands.

File: pkg/cmd/service/service_test.go
Comment (craigpastro) [reply pkg/cmd/service/service_test.go]: Ah, was using for debugging. Reverted. Thanks.

File: pkg/cmd/service/service_test.go
Comment (craigpastro) [reply pkg/cmd/service/service_test.go]: Moved to `ensureServiceUp`. Thanks!

File: pkg/cmd/service/service_test.go
Comment (craigpastro) [reply pkg/cmd/service/service_test.go]: Done. Thanks!

File: pkg/cmd/service/service_test.go
Comment (craigpastro) [reply pkg/cmd/service/service_test.go:681]: `canlceling` is the thing that is triggering the graceful closure. Server.Close() closes database and auth connections only. It's a bit ... odd maybe?

File: pkg/cmd/service/service_test.go
Comment (elbuo8) [inline pkg/cmd/service/service_test.go:51]: I'd always spin it up so we don't need to have the context. With the introduction of the healthcheck we ensure we don't run into a race condition. Plus the upcoming work for having random ports assigned should help too.

File: pkg/cmd/service/service_test.go
Comment (elbuo8) [inline pkg/cmd/service/service_test.go:98]: `openFGAServerAddr` has this config rather than hardcoding. That way when this gets refactored we don't have to look it up.

Might be worth also putting it in the `backoff` function after the GRPC healthcheck.

File: pkg/cmd/service/service_test.go
Comment (miparnisari) [inline pkg/cmd/service/service_test.go:97]:
```suggestion
	if checkHttp {
```

?

File: pkg/cmd/service/service_test.go
Comment (jon-whit) [reply pkg/cmd/service/service_test.go:98]: Wait, why do we have the HTTP option instead of just using the single RPC health check? I feel like this is extra code that isn't needed.

Comment (elbuo8) [reply pkg/cmd/service/service_test.go:98]: The RPC gateway might boot up while the HTTP listener might not. We want to ensure both are up.

Comment (craigpastro) [reply pkg/cmd/service/service_test.go]: What Yamil said. I suppose this would be fixed with retryable client everywhere.

Another problem however, was some tests were failing because we could cancel the context when the HTTP server was in the middle of starting up resulting in an error. This is probably fixed by disabling the HTTP server in such tests.

In any case, I think this adds robustness.

File: pkg/cmd/service/service_test.go
Comment (jon-whit) [inline pkg/cmd/service/service_test.go:51]: `config.HTTP.Addr` is the addr. For example if I set `config.HTTP.Addr: "127.0.0.1:8082"` then the HTTP server will only bind on host address 127.0.0.1 and port 8082. 

So the `localhost` component here isn't quite right.

Comment (jon-whit) [reply pkg/cmd/service/service_test.go:51]: 🤔 but you're using the default config above which will be like `:8080`. This just feels weird. What's a better way to do it? I'm thinking..

File: pkg/cmd/service/service_test.go
Comment (jon-whit) [inline pkg/cmd/service/service_test.go:222]: The reason we used `GetServiceConfig` was that it is the config entrypoint for the service. It is supposed to call `DefaultConfig` and merge other user provided configs ontop of it. 

I think that we actually get better test coverage and test fidelity if we leave this as it was before. It also gives us the apportunity to use `os.SetEnv(...)` to test that our env variables are wired up in our service properly and that `GetServiceConfig` respects it. So I think I'd prefer to leave it even though it is one extra line of code for the `require.NoError(...)`.

File: pkg/cmd/service/service_test.go
Comment (craigpastro) [reply pkg/cmd/service/service_test.go:51]: Use netip.AddrPort and make it explicit everywhere? I can do that in a following PR.

Comment (jon-whit) [reply pkg/cmd/service/service_test.go:51]: I guess I'll need to do something like this with the changes I'm making for the random port. I'll include that in that PR.

File: pkg/cmd/service/service_test.go
Comment (jon-whit) [inline pkg/cmd/service/service_test.go:103]: Can we rename this to `httpHealthCheck`?

**Review by jon-whit — DISMISSED** [review]

File: pkg/cmd/service/service_test.go
Comment (craigpastro) [reply pkg/cmd/service/service_test.go]: Done. Thanks!

File: pkg/cmd/service/service_test.go
Comment (craigpastro) [reply pkg/cmd/service/service_test.go]: Renamed. Thank you!

File: pkg/cmd/service/service_test.go
Comment (craigpastro) [reply pkg/cmd/service/service_test.go]: It makes sense. I thought that if we need it in a particular test we could use it, but I see what you are saying too. Reverted. Thanks!

**Review by jon-whit — APPROVED** [review]

**Pull request merged**
